draft-ietf-mls-protocol.md

```diff
@@ -3553,6 +3553,99 @@ struct {
+
+* It contains an Update proposal generated by the committer.
+
+* It contains a Remove proposal that removes the committer.
+
+* It contains multiple Update and/or Remove proposals that apply to the same leaf.
+  If the committer has received multiple such proposals they SHOULD prefer any Remove
+  received, or the most recent Update if there are no Removes.
+
+* It contains multiple Add proposals that contain KeyPackages that represent the same
+  client according to the application (for example, identical KeyPackages or KeyPackages
+  sharing the same Credential).
+
+* It contains multiple PreSharedKey proposals that reference the same PreSharedKeyID.
+
+* It contains multiple GroupContextExtensions proposals.
+
+* It contains a ReInit proposal together with any other proposal. If the committer has
+  received other proposals during the epoch, they SHOULD prefer them over the
+  ReInit proposal, allowing the ReInit to be resent and applied in a subsequent
+  epoch.
+
+* It contains an ExternalInit proposal.
+
+* It contains a proposal with a non-default proposal type that is not supported by some
+  members of the group that will process the Commit (i.e., members being added
+  or removed by the Commit do not need to support the proposal type).
+
+* After processing the commit the ratchet tree is invalid, in particular, if it
+  contains any leaf node that is invalid according to {{leaf-node-validation}}.
+
+An application may extend the above procedure by additional rules, for example,
+requiring application-level permissions to add members, or rules concerning
+non-default proposal types.
+
+For an external commit, the list is valid if it contains only the following proposals
+(not necessarily in this order):
+
+* Exactly one ExternalInit
+
+* At most one Remove proposal, with which the joiner removes an
+  old version of themselves. If a Remove proposal is present, then the LeafNode in the
+  `path` field of the external commit MUST meet the same criteria as would the LeafNode
+  in an Update for the removed leaf (see {{update}}). In particular, the `credential`
+  in the LeafNode MUST present a set of identifiers that is acceptable to the
+  application for the removed participant.
+
+* Zero or more PreSharedKey proposals.
+
+* No other proposals.
+
+Proposal type defined in the future may make updates to the above validation
+logic to incorporate considerations related to proposals of the new type.
+
+## Applying a Proposal List
+
+The sections above defining each proposal type describe how each individual
+proposals are applied.  When creating or processing a Commit, a client applies a
+list of proposals to the ratchet tree and GroupContext. The client MUST apply
+the proposals in the list in the following order:
+
+* If there is a GroupContextExtensions proposal, replace the `extensions` field
+  of the GroupContext for the group with the contents of the extension.
```

Comment:
Interim 2022-06-02:

* Clarify that this means that the new extensions apply to other proposals in the same list.